ykbh/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+            _ => unreachable!()
+        };
+    }
+
+    fn store_raw(&self, dest: usize, src: Object) {
+        unsafe {
+            std::ptr::copy(src.data.as_ptr(), dest as *mut u8, src.data.len());
+        }
+    }
+
+    fn to_obj(&self, src: &IPlace) -> Object {
+        match src {
+            IPlace::Val { local, off, ty } => {
+                if *off == 0 {
+                    // If the offset is 0 we can just copy the object, e.g. $1 = $2.
+                    if let Some(obj) = self.get_var(local) {
+                        obj.clone()
+                    } else {
+                        unreachable!()
+                    }
+                } else {
+                    // Copy data from an offset, e.g. $1 = $2+2.
+                    if let Some(obj) = self.get_var(local) {
+                        let ptr = obj.data.as_ptr() as usize + *off as usize;
+                        let size = SIR.ty(ty).size();
+                        Object::from_ptr(ptr, size as usize)
+                    } else {
+                        unreachable!()
+                    }
+                }
+            }
+            IPlace::Indirect { ptr, off, ty } => {
+                // FIXME copy only depending on type?
+                // Dereference a pointer and copy the data it points to.
+                if let Some(obj) = self.get_var(&ptr.local) {
+                    let dstptr = obj.to_usize() + ptr.off as usize + *off as usize;
+                    let size = SIR.ty(ty).size();
+                    Object::from_ptr(dstptr, size as usize)
+                } else {
+                    unreachable!()
+                }
+            }
+            IPlace::Const { val, ty: _ty } => {
+                // The source is a constant: create a new object.
+                match val {
+                    Constant::Int(ci) => match ci {
+                        ConstantInt::UnsignedInt(ui) => {
+                            match ui {
+                                UnsignedInt::U8(v) => Object::from_u8(*v),
+                                _ => todo!()
+                            }
+                        }
+                        ConstantInt::SignedInt(_si) => todo!(),
+                    },
+                    Constant::Bool(_b) => todo!(),
+                    Constant::Tuple(t) => {
+                        if SIR.ty(t).size() == 0 {
+                            ZST
+                        } else {
+                            todo!()
+                        }
+                    },
+                    _ => todo!(),
+                }
+            }
+            _ => todo!()
+        }
+    }
+
+
+    /// Creates a reference to an IPlace.
+    fn mkref(&mut self, dest: &IPlace, src: &IPlace) {
+        let obj = match src {
+            IPlace::Val { local, off, ty: _ty } => {
+                if let Some(obj) = self.get_var(local) {
+                    // The pointer to the vector should remain valid, since no changes to the
```

Comment:
Ah. This is why you can't represent local storage as a slice into one big vector then. That vector will grow as we add new frames and the buffer will move around.

You could store a slice as an `index, len` pair though. That'd be immune.

Reply:
But then you'd have to treat pointers given to us from outside specially. Argh!

Reply:
I don't think I'm a fan of the growing big vector. Though we could have once big vector per frame. That should work.

Reply:
> But then you'd have to treat pointers given to us from outside specially. Argh!

Yeah, I think my first attempt will be to replace the vecs with arrays and see how far I get with that.

Reply:
> Though we could have once big vector per frame

sounds good.